src/cohort/cohort_server.cc

```diff
@@ -1,56 +1,265 @@
+  common::AbortReason abort_reason;
+  if (abort_status.has_value()) {
+    // It's okay if it fails since it will auto-abort at the presumed abort
+    // time.
+    Blockchain::Vote(transaction_id, cohort_index, Blockchain::Ballot::ABORT)
+        .IgnoreError();
+    switch (abort_status->code()) {
+      case absl::StatusCode::kDeadlineExceeded:
+        abort_reason = common::ABORT_REASON_PRESUMED_ABORT_TIMESTAMP_REACHED;
+        break;
+      case absl::StatusCode::kNotFound:
+        abort_reason = common::ABORT_REASON_OPERATION_FOR_NON_EXISTENT_VALUE;
+        break;
+      default:
+        abort_reason = common::ABORT_REASON_UNSPECIFIED;
+    }
+  }
+  metadata_by_transaction_id_[transaction_id].response.set_aborted_response(
+      abort_reason);
+  // TODO(benjmarks22): Maybe add retry logic here?
+  metadata_by_transaction_id_[transaction_id].db->Abort().IgnoreError();
+  CleanUpTransactionMetadata(transaction_id);
+}
+
+void CohortServer::CommitTransaction(const std::string& transaction_id) {
+  // The database must commit the transaction at this point, so the only
+  // acceptable failures are transient ones (e.g. deadline exceeded). Thus we
+  // retry until it succeeds.
+  while (!metadata_by_transaction_id_[transaction_id].db->Commit().ok()) {
+  }
+  // This is necessary if it's a write only transaction to ensure the response
+  // indicates that it committed.
+  metadata_by_transaction_id_[transaction_id]
+      .response.mutable_committed_response();
+  CleanUpTransactionMetadata(transaction_id);
+}
+
+absl::Status CohortServer::PersistTransactionResults(
+    const std::string& transaction_id) {
+  const std::string& path =
+      absl::StrCat(db_txn_response_dir_, "/", transaction_id, ".binarypb");
+  std::fstream output(path, std::ios::out | std::ios::trunc | std::ios::binary);
+  if (!metadata_by_transaction_id_[transaction_id]
+           .response.committed_response()
+           .SerializeToOstream(&output)) {
+    return absl::InternalError(
+        "Failed to write transaction responses to disk.");
+  }
+  return absl::OkStatus();
+}
+
+void CohortServer::ProcessTransaction(
+    const PrepareTransactionRequest& request) {
+  internal::TransactionMetadata& metadata =
+      metadata_by_transaction_id_[request.transaction_id()];
+  metadata.response.mutable_pending_response();
+  metadata.db = db_transaction_adapter_creator_();
+  const absl::Time presumed_abort_time =
+      absl::FromUnixSeconds(request.config().presumed_abort_time().seconds()) +
+      absl::Nanoseconds(request.config().presumed_abort_time().nanos());
+
+  const absl::Status db_status = ProcessTransactionInDb(
+      request.transaction(), presumed_abort_time, metadata);
+  if (!db_status.ok()) {
+    AbortTransaction(request.transaction_id(), request.cohort_index(),
+                     db_status);
+    return;
+  }
+
+  const absl::Status persist_status =
+      PersistTransactionResults(request.transaction_id());
```

Comment:
Should we call `PersistTransactionResults` before `ProcessTransactionInDb` (write ahead logging)? With that, a recovered cohort instance can always replay the changes to database; on the other hand, with the current implementation, we might process the transaction in database then crash before writing to log.

Reply:
If we crash before writing to log, then we'll just abort. The database should not actually be updating until the `Commit` method is called. I think this is just bad naming of the methods since `ProcessTransactionInDb` doesn't actually update the database. It just makes it ready so that the database can commit it later. I'll see if I can rename the methods here to be more clear, but I'm open to ideas.

Reply:
I see. So, it only matters if we've written to log before voting yes so we can process (prepare) the transaction again on recovery.

Reply:
I changed it to log both the request and responses. The responses are useful if the DB commits the transaction but the cohort crashes before the client is done requesting the get responses. The request is useful if the DB crashes or the cohort crashes after voting to commit and before the DB commits the transaction.